File: 64-minimum-path-sum/64-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int r = grid.length;
        int c = grid[0].length;
        //System.out.println("r= " + r + " , c= " + c );
        int[][] dp = new int[r][c];
        int t = 0;
        dp[0][0] = grid[0][0];
        for(int i = 1; i < c; i++){
            dp[0][i] = grid[0][i] + dp[0][i-1];
            //System.out.print(dp[0][i] + " ");
        }
        
        //System.out.println();
        
        for(int i = 1; i < r; i++){
            dp[i][0] = grid[i][0] + dp[i-1][0];
            //System.out.print(dp[i][0] + " ");
        }
        
        for(int i = 1; i < r; i++){
            for(int j = 1; j < c; j++){
                dp[i][j] = grid[i][j] + Math.min(dp[i-1][j], dp[i][j-1]);
            }
        }
        
        
        return dp[r-1][c-1];
        
    }
}

File: 64-minimum-path-sum/64-minimum-path-sum.cpp
class Solution {
public:
//     int helper(int i, int j, vector<vector<int>>& grid){
//         if(i == 0 && j == 0) return grid[0][0];
//         if(i < 0 || j < 0) return 1e6;
        
//         int up = grid[i][j] + helper(i-1, j, grid);
//         int left = grid[i][j] + helper(i, j-1, grid);
        
//         return min(up, left);
//     }
    
    //this is space optimised and the previous one is non space optimised
    int minPathSum(vector<vector<int>>& grid) {
        int r = grid.size(), c = grid[0].size();
        
        vector<int> prev(c, 0);
        
        for(int i = 0 ; i < r; i++){
            vector<int> temp(c, 0);
            for(int j = 0; j < c; j++){
                if(i == 0 && j == 0) temp[j] = grid[0][0];
                else{
                    //important thing is initialise up and left with maximum number as we are finding min path sum
                    int up = 1e6, left = 1e6;
                    if(i > 0) up = grid[i][j] + prev[j];
                    if(j > 0) left = grid[i][j] + temp[j-1];
                    
                    temp[j] = min(up, left);
                }
            }
            
            prev = temp;
        }
        
        return prev[c-1];
        
        // return helper(r-1, c-1, grid);
    }
};